### PR TITLE
Transform config command into a namespace

### LINKED ISSUE #149: Improve configuration and default values
Hello,

I've been thinking about how to improve the way Gandi CLI allows users to
get/set configuration and default values.

This is what I took into consideration:
- There is no gandi command to display current configuration and/or default values.
- There's no way to display single specific configuration and/or defaults values.
- Default values are overriden at every command

This is what I would suggest, freely inspired from NPM CLI :

- gandi config
  - gandi config set ```<key> <value>``` [-g]
  - gandi config get ```[<key>]``` [-g]
  - gandi config delete ```<key>```  [-g]
  - gandi config list  [-g]
  - gandi config edit  [-g]

- set a default value in config.yml only if the key is not present yet.

As per current behavior, -g should force use of global configuration for the given action.

I'm willing to open an MR and work on this.

What do you think?

Related to #91 and #54

### PR DESCRIPTION
**Pull request by ziir**
Configuration and defaults management made more flexible.
All `gandi config` commands accept the `-g` option to only
target the global configuration.

- [x] gandi config
  - [x] gandi config set ```<key> <value>``` [-g]
  - [x] gandi config get ```[<key>]``` [-g]
  - [x] gandi config delete ```<key>```  [-g]
  - [x] gandi config list  [-g]
  - [x] gandi config edit  [-g]

- [x] set a default value in configuration file only if the key is not present yet.

Fixes #149 and #54

### CODE REVIEW

File: gandi/cli/commands/config.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+@cli.command()
+@click.option('-g', help='Edit global configuration (default=local).',
+              is_flag=True, default=False)
+@pass_gandi
+def edit(gandi, g):
+    """Edit config file with prefered text editor"""
+    config_file = gandi.home_config if g else gandi.local_config
+    path = os.path.expanduser(config_file)
+    editor = gandi.get('editor', '/usr/bin/vi')
```
Comment (aegiap) [inline gandi/cli/commands/config.py]: I think we should avoid hardcoding path. Why not default to 'vim' ?
The $EDITOR environment variable from the user shell seems to be common, we should be using it.

File: gandi/cli/core/conf.py [hunk 2]
```diff
@@ -116,6 +117,19 @@ def _del(cls, scope, key, separator='.', conf=None):
                 return
 
     @classmethod
+    def delete(cls, global_, key):
+        """ Delete key/value pair from configuration file. """
+        # first retrieve current configuration
+        scope = 'global' if global_ else 'local'
+        config = cls._conffiles.get(scope, {})
+        if config:
+            # apply modification to fields
```
Comment (MoiTux) [inline gandi/cli/core/conf.py]: I'm not the comment is relevant here

File: gandi/cli/core/conf.py [hunk 3]
```diff
@@ -150,7 +164,7 @@ def _get(cls, scope, key, default=None, separator='.'):
             return default
 
     @classmethod
-    def get(cls, key, default=None, separator='.'):
+    def get(cls, key, default=None, separator='.', global_=False):
         """ Retrieve a key value from loaded configuration.
 
         Order of search :
```
Comment (MoiTux) [inline gandi/cli/core/conf.py]: that's is not always true anymore

File: gandi/cli/commands/config.py (hunk 1 above)
Comment (bapt) [inline gandi/cli/commands/config.py]: Hardcoding the path of vi is wrong, some OS (used to?) have vi in a different path, like old redhat had vi in /bin don't know if they still do. In C you have a macro for that _PATH_VI defined in paths.h which of course is not available in python so I would go for a plain 'vi' and rely on the path to find the proper path.
Defaulting on 'vim' won't imho be a good idea, has you have the insurance that vi is installed on anything trying to respect POSIX (part of POSIX.1-2008) (ok I know no one care about posix :) but hey)
In anycase +1 on adding support for the EDITOR variable

File: gandi/cli/core/conf.py [hunk 4]
```diff
@@ -187,6 +202,15 @@ def configure(cls, global_, key, val):
         cls.save(os.path.expanduser(conf_file), config)
 
     @classmethod
+    def list(cls, global_):
+        """ Return configuration file content. """
+        cls.load_config()
```
Comment (MoiTux) [inline gandi/cli/core/conf.py]: for me it's useless the decorator `pass_gandi` do that for you

File: gandi/cli/commands/config.py (hunk 1 above)
Comment (MoiTux) [inline gandi/cli/commands/config.py]: I don't know for all but git seems to default to the config and falls back to the environment variable maybe we should do the same ?  
https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup#Your-Editor

Comment (bapt) [inline gandi/cli/commands/config.py]: Yes that is exactly what should be done:
Lookup the config if nothing lookup the env for EDITOR is nothing vi

Comment (ziir) [inline gandi/cli/commands/config.py]: Thanks for the feedback.
The $EDITOR environment variable is already supported as 
`gandi get editor` or `gandi.get('editor', 'vi')` will :
- Check the $EDITOR env variable (the key, capitalized)
- Check the local config file
- Check the global config file
- Use the default

There is a test in `test_config.py` that ensures it works as intended.

I'll use `vi` as default instead of `/usr/bin/vi`.

File: gandi/cli/core/conf.py (hunk 4 above)
Comment (ziir) [inline gandi/cli/core/conf.py]: Thanks, I'll take a look

File: gandi/cli/commands/config.py (hunk 1 above)
Comment (ziir) [inline gandi/cli/commands/config.py]: >  Lookup the config if nothing lookup the env for EDITOR is nothing vi
Problem is that it's exactly the opposite of what `gandi.get()` does :/

Would you agree with the following suggestion, or do you prefer the current behavior:
- Check global configuration only (not local)
- Fallback to $EDITOR
- Fallback to 'vi'

Comment (bapt) [inline gandi/cli/commands/config.py]: skipping the local is imho a wrong idea, so it would be ok with the current behavior of ${EDITOR} then local then global then fallback.

Comment (ziir) [inline gandi/cli/commands/config.py]: Replaced fallback to 'vi' and kept current behavior.

File: gandi/cli/core/conf.py (hunk 2 above)
Comment (ziir) [inline gandi/cli/core/conf.py]: Updated

File: gandi/cli/core/conf.py (hunk 3 above)
Comment (ziir) [inline gandi/cli/core/conf.py]: Updated

File: gandi/cli/core/conf.py (hunk 4 above)
Comment (ziir) [inline gandi/cli/core/conf.py]: Useless indeed, removed.

File: gandi/cli/commands/config.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+@cli.command()
+@click.option('-g', help='Edit global configuration (default=local).',
+              is_flag=True, default=False)
+@pass_gandi
+def edit(gandi, g):
+    """Edit config file with prefered text editor"""
+    config_file = gandi.home_config if g else gandi.local_config
+    path = os.path.expanduser(config_file)
+    editor = gandi.get('editor', 'vi')
```
Comment (mardiros) [inline gandi/cli/commands/config.py]: Something like that ?

```
editor = gandi.get('editor', os.environ.get('EDITOR'))
while not editor:
    # raw_input in py2 / input in py3, maybe a helper exists for that in the project
    editor = raw_input('editor is not configured, please enter the path of your editor:')
```

Comment (ziir) [inline gandi/cli/commands/config.py]: gandi.get('editor') already checks the environment variable.
Are you saying you'd prefer asking the user for his prefered editor rather than using `vi` as default?

(See previous comments on outdated diff for discussion about this)

File: gandi/cli/core/conf.py [hunk 6]
```diff
@@ -187,6 +203,14 @@ def configure(cls, global_, key, val):
         cls.save(os.path.expanduser(conf_file), config)
 
     @classmethod
+    def list(cls, global_):
+        """ Return configuration file content. """
+        scope = 'global' if global_ else 'local'
+        config = cls._conffiles.get(scope, {})
+        if config:
```
Comment (MoiTux) [inline gandi/cli/core/conf.py]: This is always True due to the get(scope, {})

File: gandi/cli/tests/commands/test_config.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+class ConfigTestCase(CommandTestCase):
+
+    def test_get_empty(self):
+        result = self.invoke_with_exceptions(config.get, [])
+        self.assertEqual(result.exit_code, 2)
+
+    def test_get_unknown(self):
+        result = self.invoke_with_exceptions(config.get, ['unknown-key'])
+        self.assertEqual(result.output, """No value found.
```
Comment (mardiros) [inline gandi/cli/tests/commands/test_config.py]: Too me, it is more readable/explicit with :

```
       self.assertEqual(result.output, 'No value found.\n')
```

Comment (ziir) [inline gandi/cli/tests/commands/test_config.py]: I took example on the other tests, all of them use the same notation.

File: gandi/cli/core/conf.py (hunk 6 above)
Comment (MoiTux) [inline gandi/cli/core/conf.py]: You should always return the dirct even if it's empty, this give the possibility to call the function in a for loop without any error.  
so `return cls._conffiles.get(scope, {})`

File: gandi/cli/commands/config.py (hunk 5 above)
Comment (mardiros) [inline gandi/cli/commands/config.py]: If the env var is checked, this is great.

Using vi to "simplify" sounds wrong to me.

If you never used vi you will have no idea of how to edit the file.
Some editors like nano are more friendly for newbies, but some people
do not like it too. So asking for the user to set the EDITOR looks great to me.

Comment (mardiros) [inline gandi/cli/commands/config.py]: notepad.exe

File: gandi/cli/tests/commands/test_config.py (hunk 7 above)
Comment (MoiTux) [inline gandi/cli/tests/commands/test_config.py]: I agree with @mardiros but maybe it's better to be consistent ?

File: gandi/cli/tests/commands/test_contact.py [hunk 8]
```diff
@@ -152,5 +152,7 @@ def test_create_apikey_ok(self):
 Will save your apikey into the config file.""")
         self.assertEqual(result.exit_code, 0)
 
-        self.assertEqual(GandiModule._conffiles,
```
Comment (sayoun) [inline gandi/cli/tests/commands/test_contact.py]: why change this test ? It should still work as before your new feature

File: gandi/cli/tests/commands/test_config.py (hunk 7 above)
Comment (sayoun) [inline gandi/cli/tests/commands/test_config.py]: You should be consistent and keep the same notation as in other tests, so it doesn't disturb people looking in the code the first time.

File: gandi/cli/commands/config.py (hunk 5 above)
Comment (sayoun) [inline gandi/cli/commands/config.py]: You should check for existing editor value in the env var (through the gandi.get method indeed) but never store this value in the CLI configuration file as this value doesn't belong here.

If the value is not present, you should make a click.prompt call to retrieve a value but if not provided just exit with a nice warning message like editor not configured or not found.

File: gandi/cli/tests/commands/test_contact.py (hunk 8 above)
Comment (ziir) [inline gandi/cli/tests/commands/test_contact.py]: In previous implementation, I used to create a default `editor` config value, and had to update this test.
I then understood how wrong I was, updated the code, but forgot to revert the change to this test.
Doing it now.

File: gandi/cli/commands/config.py (hunk 5 above)
Comment (ziir) [inline gandi/cli/commands/config.py]: > but never store this value in the CLI configuration file as this value doesn't belong here.

Yep, I'm not storing it anywhere from here.
I still should give the possibility for the user to do so via `gandi config set editor /usr/bin/vi`, right?

I'll update the code for the prompt + warning / exit.

Thank you.

File: gandi/cli/commands/config.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import subprocess
+
+import click
+
+from gandi.cli.core.cli import cli
+from gandi.cli.core.params import pass_gandi
+
+
+@cli.command()
+@click.option('-g', help='Get from global configuration only'
```
Comment (yanndinendal) [inline gandi/cli/commands/config.py]: Missing space?

`'Get from global configuration only '`

File: gandi/cli/tests/commands/test_contact.py (hunk 8 above)
Comment (ziir) [inline gandi/cli/tests/commands/test_contact.py]: Though I find this test a little annoying the way it is right now, because it always fails if you have a local configuration file. I think the way I had changed it avoided that while keeping the test's objective.

Comment (sayoun) [inline gandi/cli/tests/commands/test_contact.py]: I never tested with a local configuration file, as I use tox or travis to run tests, but your point is valid indeed, if it can avoid a failed test you can keep your changes, sorry to answer late as I see you already have reverted the change

File: gandi/cli/core/params.py [hunk 10]
```diff
@@ -503,15 +503,23 @@ def consume_value(self, ctx, opts):
         return value
 
     def handle_parse_result(self, ctx, opts, args):
-        """ Save value for this option in configuration. """
+        """ Save value for this option in configuration
+        if key/value pair doesn't already exist.
+        Or old value in config was deprecated
+        it needs to be updated to the new value format
+        but the value keeps the same "meaning"
+        """
```
Comment (ziir) [inline gandi/cli/core/params.py]: Tests are work in progress.

File: gandi/cli/tests/commands/test_contact.py (hunk 8 above)
Comment (ziir) [inline gandi/cli/tests/commands/test_contact.py]: Aaaand it's back

**Comment by sayoun** [comment]
Seems good to me, I'll merge this.
